esp-hal-embassy/src/time_driver.rs

```diff
@@ -1,85 +1,130 @@
-use core::cell::{Cell, RefCell};
+use core::cell::Cell;
 
-use critical_section::Mutex;
 use embassy_time_driver::{AlarmHandle, Driver};
 use esp_hal::{
     interrupt::{InterruptHandler, Priority},
     prelude::*,
+    sync::Locked,
     time::now,
     timer::{AnyTimer, OneShotTimer},
 };
 
-pub const MAX_SUPPORTED_ALARM_COUNT: usize = 7;
-
 pub type Timer = OneShotTimer<'static, AnyTimer>;
 
-static TIMERS: Mutex<RefCell<Option<&'static mut [Timer]>>> = Mutex::new(RefCell::new(None));
+enum AlarmState {
+    Created(extern "C" fn()),
+    Allocated(extern "C" fn()),
+    Initialized(&'static mut Timer),
+}
 
-#[allow(clippy::type_complexity)]
-struct AlarmState {
-    pub callback: Cell<Option<(fn(*mut ()), *mut ())>>,
-    pub allocated: Cell<bool>,
+struct Alarm {
+    pub callback: Cell<(*const (), *mut ())>,
+    pub state: AlarmState,
 }
 
-unsafe impl Send for AlarmState {}
+unsafe impl Send for Alarm {}
 
-impl AlarmState {
-    pub const fn new() -> Self {
+impl Alarm {
+    pub const fn new(handler: extern "C" fn()) -> Self {
         Self {
-            callback: Cell::new(None),
-            allocated: Cell::new(false),
+            callback: Cell::new((core::ptr::null(), core::ptr::null_mut())),
+            state: AlarmState::Created(handler),
         }
     }
 }
 
 pub(super) struct EmbassyTimer {
-    alarms: Mutex<[AlarmState; MAX_SUPPORTED_ALARM_COUNT]>,
+    alarms: Locked<[Alarm; MAX_SUPPORTED_ALARM_COUNT]>,
```

Comment:
This comment is beyond the goal of this PR but I thought I'd mention it, in case it tickles your fancy.

A lock per alarm (i.e. `[Locked<Alarm>; MAX_SUPPORTED_ALARM_COUNT]>`) would mean even less contention than a lock for all alarms (i.e. `Locked<[Alarm; MAX_SUPPORTED_ALARM_COUNT]>`)

Reply:
> This comment is beyond the goal of this PR but I thought I'd mention it, in case it tickles your fancy.

No it isn't and I guess you're right.